proposals/csharp-9.0/nullable-reference-types-specification.md

```diff
@@ -285,62 +307,155 @@ If a `simple_name` is not classified as a value, its null state is "not null". O
+    // Warning: Dereference of a possibly null reference.
+    Console.WriteLine(o.ToString());
+}
+
+// Nullable friendly pattern
+if (array[0] is {} o)
+{
+    Console.WriteLine(o.ToString());
+}
+```
+
 ### Base access
 
 If `B` denotes the base type of the enclosing type, `base.I` has the same null state as `((B)this).I` and `base[E]` has the same null state as `((B)this)[E]`.
 
 ### Default expressions
 
-`default(T)` has the null state "not null" if `T` is known to be a nonnullable value type. Otherwise it has the null state "maybe default".
+`default(T)` has the null state based on the properties of the type `T`:
+
+- If the type is a *nonnullable* type then it has the null state "not null"
+- Else if the type is a type parameter then it has the null state "maybe default"
+- Else it has the null state "maybe null"
 
 ### Null-conditional expressions
 
-A `null_conditional_expression` has the null state "maybe null".
+A `null_conditional_expression` has the null state based on the properties of expression type. Note that this refers to the type of the `null_conditional_expression`, not the original type of the member being invoked:
+
+- If the type is a *nullable* value type then it has the null state "maybe null"
+- Else if the type is a *nullable* type parameter then it has the null state "maybe default"
+- Else it has the null state "maybe null"
 
 ### Cast expressions
 
-If a cast expression `(T)E` invokes a user-defined conversion, then the null state of the expression is the default null state for its type. Otherwise, if `T` is *nullable* then the null state is "maybe null". Otherwise the null state is the same as the null state of `E`.
+If a cast expression `(T)E` invokes a user-defined conversion, then the null state of the expression is the default null state for the type of the user-defined conversion. Otherwise:
+
+- If `T` is a *nullable* type in the form `U?` where `U` is a type parameter then the null state is "maybe default"
+- Else If `T` is a *nullable* type then the null state is "maybe null"
+- Else the null state is the same as the null state of `E`
+
+### Unary and binary operators
 
-***This needs upddating***
+If a unary or binary operator invokes an user-defined operator then the null state of the expression is the default null state for the type of the user-defined operator. Otherwise it is the null state of the expression.
+
+***Something special to do for binary `+` over strings and delegates?***
 
 ### Await expressions
 
 The null state of `await E` is the default null state of its type.
 
 ### The `as` operator
 
-An `as` expression has the null state "maybe null".
+The null state of an `E as T` expression depends first on properties of the type `T`. If the type *nonnullable* then the null state is "not null". Otherwise the null state depends on the conversion from the type of `E` to type `T`:
+
+- If the conversion Identity, boxing, implicit reference or implicit nullable has the same null state as `E`
```

Comment:
>If the conversion Identity, boxing, implicit reference or implicit nullable has the same null state as `E` [](start = 2, length = 106)

Perhaps "If the conversion is an identity, implicit reference, or implicit nullable conversion, then the null state is the null state of `E`"